add events task

Fln1k opened the pull request.

Fln1k committed:
initital commit

Guardian passwordless authentication
Bamboo mailer
Add user and events

Fln1k committed:
add event_confirmation

Fln1k committed:
add email change modal window

Fln1k committed:
fix logout bug

Fln1k committed:
automatically confirmation for event author

Fln1k commented:
need to add your username and password for gmail.com in events_task/config/config.exs

azhi commented on events_task/lib/my_app.ex:
```diff
@@ -0,0 +1,9 @@
+defmodule MyApp do
```
Not very important, but try to name things semantically.
Here root module is called `MyApp` - i'm pretty sure this is example name :)
`EventsTask` would be a better one.
No need to rename it everywhere, just a general note.

azhi commented on events_task/lib/my_app/accounts/accounts.ex:
```diff
@@ -0,0 +1,44 @@
+  end
+
+  def get_or_create_by_email(email) do
+    case Repo.get_by(User, email: email) do
```
This check will work for single-threaded application, but elixir is multi-threaded by default.
What will eventually happen is:
* user will double-click on register button, two requests to HTTP server will be issued, two elixir threads will be created to process them
* first thread will reach this line, issue a request to DB, see that there are no records, and will be paused by CPU awaiting next execution cycle
* second thread will reach this line, issue a request to DB, see that there are no records, and will continue and create new user
* first thread will be woken up, will continue and create new user
The result is that we will have two users with same email in DB.

To 100% avoid this issue, DB has unique indexes - whenever you don't want some duplicate rows exist, you should create one.
Ecto has a feature `Ecto.Changeset.unique_constraint/2` - with this feature, you don't need to manually check whether record exist or not - ecto will try to insert it, and if it will encounter an unique constraint validation - it will parse it and convert to validation error.

azhi commented on events_task/lib/my_app/content/content.ex:
```diff
@@ -0,0 +1,108 @@
+
+  """
+  def create_event(attrs \\ %{}) do
+    IO.inspect("---------------------------------------------------------")
```
Try not to commit debug output.
Oh, and you can debug with:
* `Logger.debug` (lowering logger level to debug) - best option if you want to make sure you won't cause issues on production system
* `Logger.info` - still more useful that just `IO.puts` - writes nice message with timestamp, colorized depending on level, with some protection from multi-threaded output that intersects with each other.
* `IEx.pry` - `binding.pry` from ruby, launches an interactive iex session at this point in code so you can poke around, see values of local variables, etc.

azhi commented on events_task/lib/my_app/email.ex:
```diff
@@ -0,0 +1,11 @@
+defmodule MyApp.Email do
+  use Bamboo.Phoenix, view: MyApp.EmailView
+
+  def welcome_text_email(email_address, link) do
```
If I understand this correctly, this email is sent on every login.
If it's true, then it's another lower importance naming issue - I'd say `welcome` is something that is sent once, and it should be called `login_link_email` or something like that.

azhi commented on events_task/lib/my_app/user_event_confirmation/user_event_confirmation.ex:
```diff
@@ -0,0 +1,126 @@
+    Repo.all(Confirmation)
+  end
+
+  def user_list_events(user) do
```
Another naming issue - this func doesn't return 'all' user events, and it actually doesn't return 'events' :).
Better name would be `list_user_confirmed_event_ids`.

azhi commented on events_task/lib/my_app/user_event_confirmation/user_event_confirmation.ex:
```diff
@@ -0,0 +1,126 @@
+
+  def user_list_events(user) do
+    if user do
+      Repo.all(
```
To avoid unnecessary nesting, it's common practice to pipe queries into repo:
```
from(u in Confirmation,
  where: u.user_id == type(^user.id, :integer),
  select: u.event_id
)
|> Repo.all()
```

azhi commented on events_task/lib/my_app/user_event_confirmation/user_event_confirmation.ex:
```diff
@@ -0,0 +1,126 @@
+    if user do
+      Repo.all(
+        from(u in Confirmation,
+          where: u.user_id == type(^user.id, :integer),
```
I'm not sure `type()` is needed here, ecto should be smart enough to work with `u.user_id == ^user.id`

azhi commented on events_task/lib/my_app/user_event_confirmation/user_event_confirmation.ex:
```diff
@@ -0,0 +1,126 @@
+
+    Repo.all(
+      from(u in User,
+        where: u.id in ^user_ids,
```
Whenever you can, it's best to express your DB fetches using single DB request - it avoid unnecessary overhead of making two DB queries and transforming result of first query into elixir data structures.
Here you could do join of users with confirmations:
```
from(u in User,
  join: c in assoc(u, :confirmations),
  where: u.event_id == event_id,
  select: u.email
)
|> Repo.all()
```

azhi commented on events_task/lib/my_app_web/router.ex:
```diff
@@ -0,0 +1,31 @@
+  scope "/", MyAppWeb do
+    pipe_through(:browser)
+    resources("/events", EventController)
+    get("/confirm", ConfirmationController, :create)
```
Get requests should be used only for idempotent requests (in general - requests that are not changing state of server/DB).
It's because they can be cached, and most of the tools treat them as requests that can be safely repeated.
So here confirming should be post or put.

azhi commented on events_task/lib/my_app_web/templates/event/form.html.eex:
```diff
@@ -0,0 +1,19 @@
+<%= form_for @changeset, @action, fn f -> %>
+  <%= if @changeset.action do %>
```
If you wanted to show a text when there are some errors - it would be better to test for exactly it:
`if Enum.any?(@changeset.errors)`.

Changeset having an action is only a side-effect, in general it's better not to rely on side-effects when you can check something directly.

azhi commented on events_task/lib/my_app_web/templates/event/form.html.eex:
```diff
@@ -0,0 +1,19 @@
+  <%= label f, :timestamp %>
+  <%= datetime_select f, :timestamp %>
+  <%= error_tag f, :timestamp %>
+  <%= hidden_input f,:author_id, value: @current_user.id %>
```
This works for regular uses, but this is a major security issue - value in this field is not protected at all.
As a user, I can replace value of this hidden input to any id, post form - and I will create an event with another user as author.
To fix this, `author_id` should not be set on form - it should be set in controller.

azhi commented on events_task/lib/my_app_web/templates/event/show.html.eex:
```diff
@@ -0,0 +1,31 @@
+</ul>
+<span><%= link "Back", to: "/" %></span>
+<%= if @current_user do %>
+	<%= if @current_user.id == @event.author_id do %>
```
Same as above - you are putting security checks only into templates. They should be here not to show unnecessary elements that uses can't click on, but they should be duplicated in controllers - right now any non-author user, even if he doesn't see 'Edit' button, can add `/edit` to url and successfully edit event that does not belong to him.

azhi commented on events_task/mix.exs:
```diff
@@ -0,0 +1,63 @@
+      {:phoenix_live_reload, "~> 1.0", only: :dev},
+      {:gettext, "~> 0.11"},
+      {:cowboy, "~> 1.0"},
+      {:bamboo, "~> 0.7"},
```
Try to use latest version if possible - for example `bamboo` has version `1.3.0`, which has a nice `LocalAdapter` feature that allows to easily test emails on localhost without the need to setup gmail creds :).
Latest version can be found at [https://hex.pm](https://hex.pm)

azhi commented on events_task/priv/repo/migrations/20191212174808_create_events.exs:
```diff
@@ -0,0 +1,15 @@
+    create table(:events) do
+      add(:description, :string)
+      add(:timestamp, :naive_datetime)
+      add(:author_id, references(:users, on_delete: :nothing))
```
`on_delete: :nothing` is always a bad choice because it doesn't help you keep DB data consistent.
You should always choose from one of other options:
* `:delete_all` if you want all user events to be deleted when a user is deleted
* `:nilify_all` if you events can exist without an author
* `:restrict` if you don't want to allow deleting a user when it has some events

azhi commented on events_task/lib/my_app_web/templates/event/show.html.eex:
```diff
@@ -0,0 +1,31 @@
+	<%= if @current_user.id == @event.author_id do %>
+	<span><%= link "Edit", to: event_path(@conn, :edit, @event) %></span>
+	<% end %>
+	<%= if Enum.member?(@user_email_list, @current_user.email) do %>
```
Same as security issues, except not a security issue - but still, nothing stops user from creating two confirmations. In fact, you can even do it in browser - go to event show page, click confirm, go back using browser back button, click confirm, ...
This should be fixed using unique indexes like it was described for `user.email` - this is the only way to guarantee that user can't confirm same event twice.

azhi reviewed (commented):
General issues:
* we are using CI that checks for warnings, and you have some - it's better to have a habit of not leaving warnings behind
* it's better to setup [belongs_to](https://hexdocs.pm/ecto/Ecto.Schema.html#belongs_to/3) [has_many](https://hexdocs.pm/ecto/Ecto.Schema.html#has_many/3) [has_one](https://hexdocs.pm/ecto/Ecto.Schema.html#has_one/3) associations in your `Ecto.Schema`s - they allow using ecto helpers like `assoc` in queries, which are quite useful.

Fln1k committed:
add associations & code formatting

Fln1k committed:
add security_check for edit events page

Fln1k committed:
extra features removed

Fln1k committed:
bug fix

Fln1k committed:
fix confiramtion creation method when click confirm

Fln1k commented:
The problem of POST requests for creating confirmations has not been completely resolved. After many hours of traveling around the global network, I came to the conclusion that the conn.methods property is responsible for query methods, so I came to the conclusion that I can easily change it, but this idea did not bring any result. after several hundred attempts to access the all-knowing google, I also realized that such a wonderful thing as a link always uses the Get method and one of the many ways to send a POST request is to make a form (which I did). On this, the problem of manually confirming events was resolved, thereby turning my ugly web application into something even more ugly. But the problem of automatic confirmation for the author during the creation of the event using redirect_to remained, but at the time of writing this poem my level of understanding of this problem does not allow me to formulate a correct request to the greatest source of knowledge such as Google. I just figured out how to solve the problem and I solved it within 30 seconds, but I can’t afford to delete this text so I will leave it and make the final commit

Fln1k committed:
final commit

azhi commented:
1) https://github.com/Fln1k/wellnuts_test_task/pull/5#issuecomment-568509952
2) Same as https://github.com/Fln1k/wellnuts_test_task/pull/5#discussion_r358222964 in https://github.com/Fln1k/wellnuts_test_task/pull/5/files#diff-555abc34f1fbb1b8fbcd855d628c31e2R9 - unique index is created in migration, but is not used in update
3) https://github.com/Fln1k/wellnuts_test_task/pull/5/files#diff-ee64a0e33c392d01e2b296310c96d7f4R27, https://github.com/Fln1k/wellnuts_test_task/pull/5/files#diff-5f93fd230aa464068282f95e3464c5c0R30 - underscores are reserved for unused variables, if variable is used - you should remove underscore.
4) https://github.com/Fln1k/wellnuts_test_task/pull/5/files#diff-5fbef2a23c68abf3419e60062f911a1fR21 - calling controllers from another controllers is generally a bad practice
5) https://github.com/Fln1k/wellnuts_test_task/pull/5/files#diff-5fbef2a23c68abf3419e60062f911a1fR62 - no security check
6) https://github.com/Fln1k/wellnuts_test_task/pull/5/files#r358253565

Fln1k committed:
Fix

- Combine user_event_confirmation with content.
- Add Plug for edit and update secure checks
- Fix plural confirmation bug

Fln1k committed:
Change error check

change error check on new event page

Fln1k committed:
Update event_controller.ex

fix edit error when user can't edit own events

Fln1k committed:
supplement edit&update secure check

add check on event and update functions for unauthenticated users

Fln1k committed:
fix response after failed secure check

add halt to conn conveyor in secure check to prevent responce error after failed secure check

Fln1k committed:
check redirection error

check redirection error when user tries to replace his email by  the email that already in use

azhi commented on events_task/lib/my_app_web/controllers/event_controller.ex:
```diff
@@ -0,0 +1,87 @@
+  def author_check(conn, _params) do
+    if Guardian.Plug.current_resource(conn) &&
+         Guardian.Plug.current_resource(conn).id ==
+           Content.get_event!(conn.path_params["id"]).user_id do
```
^

azhi replied:
^ Double query to DB.

azhi commented on events_task/lib/my_app/content/confirmation.ex:
```diff
@@ -0,0 +1,17 @@
+  def changeset(confirmation, attrs) do
+    confirmation
+    |> cast(attrs, [:user_id, :event_id])
+    |> unique_constraint(:user_event, name: :user_event)
```
https://hexdocs.pm/ecto/Ecto.Changeset.html#unique_constraint/3
First argument should be field.

Fln1k committed:
fix event creation for authenticated users

azhi commented on events_task/lib/my_app_web/controllers/confirmation_controller.ex:
```diff
@@ -0,0 +1,32 @@
+            |> put_flash(:info, "Confirmed successfully.")
+            |> redirect(to: "/")
+
+          {:error, %Ecto.Changeset{} = changeset} ->
```
^ here, you could now receive an error that user already confirmed event.

Fln1k committed:
add tests+small bug fix

add account test and event controller test

Fln1k committed:
add auth controller test

Fln1k committed:
add confirmation controller tset

Fln1k committed:
add content test

Fln1k committed:
add page controller test

Fln1k committed:
Delete config.exs

Fln1k committed:
change test names

Fln1k committed:
Merge branch 'events' of https://github.com/Fln1k/wellnuts_test_task into events

azhi commented on events_task/lib/my_app_web/controllers/confirmation_controller.ex:
```diff
@@ -18,9 +18,6 @@ defmodule MyAppWeb.ConfirmationController do
             |> put_flash(:info, "Confirmed successfully.")
             |> redirect(to: "/")
-
-          {:error, %Ecto.Changeset{} = changeset} ->
```
^ This was needed to catch unique error.

azhi commented on events_task/test/my_app_web/controllers/auth_controller_test.exs:
```diff
@@ -0,0 +1,98 @@
+  describe "users" do
+    test "new login page" do
+      conn = get(build_conn(), auth_path(build_conn(), :new))
+      assert conn.path_info == ["login"]
```
I'm not sure what you wanted to check here?

azhi commented on events_task/lib/my_app_web/controllers/auth_controller.ex:
```diff
@@ -13,7 +13,8 @@ defmodule MyAppWeb.AuthController do
       {:ok, user} ->
-        {:ok, _, _} = Guardian.send_magic_link(user)
+        {:ok, token, _} = Guardian.send_magic_link(user)
+        conn = assign(conn, :token, token)
```
Saving data to `conn` to only be used in tests is not encouraged, especially when it's secure-sensetive data.
I see that this is used to test magic link - better approach would be to call `Guardian.send_magic_link` in test itself.

azhi commented on events_task/test/my_app_web/controllers/auth_controller_test.exs:
```diff
@@ -0,0 +1,98 @@
+      conn = get(conn, auth_path(conn, :callback, conn.assigns[:token]))
+      conn = post(conn, auth_path(conn, :update, %{"user" => %{"email" => "Updated23mail.ru"}}))
+
+      assert conn.private[:plug_session]["phoenix_flash"] == %{
```
It's better to use helper functions from phoenix when possible ([get_flash/2](https://hexdocs.pm/phoenix/Phoenix.ConnTest.html#get_flash/2)) - phoenix guys won't break them on upgrade because they are part of public API. When you fetch stuff by hardcoded keys - nothing stops phoenix guys to rename/remove keys in next phoenix release.

azhi commented on events_task/test/my_app_web/controllers/confirmation_controller_test.exs:
```diff
@@ -0,0 +1,73 @@
+          confirmation_path(conn, :create, %{"event_id" => event.id, "user_id" => user.id})
+        )
+
+      assert conn.private[:plug_session]["phoenix_flash"] == %{
```
You can clear up old flash messages before next request using [clear_flash/1](https://hexdocs.pm/phoenix/Phoenix.ConnTest.html#clear_flash/1)

azhi commented on events_task/test/my_app_web/controllers/event_controller_test.exs:
```diff
@@ -0,0 +1,224 @@
+        |> Guardian.Plug.sign_in(create_test_user())
+        |> post(event_path(conn, :create, %{"event" => @valid_attrs}))
+
+      assert redirected_to(conn) =~ "/events/"
```
Here, you checked that controller redirected you to `/events` path - but technically nothing stops controller from redirecting to `/events` even if event was not created.
It's better to also directly check for what test description says - that some event with this attrs appeared in DB.

azhi commented on events_task/test/my_app_web/controllers/event_controller_test.exs:
```diff
@@ -0,0 +1,224 @@
+        conn
+        |> post(event_path(conn, :create, %{"event" => @valid_attrs}))
+        |> redirected_to()
+        |> String.slice(8..-1)
```
This is unreliable way to retrieve an id - path to event show page can be changed anytime, and this test will break.
Since you are not checking `:create` action here - why not just create an event using `create_test_event`?

azhi commented on events_task/test/my_app_web/controllers/event_controller_test.exs:
```diff
@@ -0,0 +1,224 @@
+          event_path(
+            conn,
+            :update,
+            struct(MyApp.Content.Event, %{
```
`struct` is used to create structs with dynamic fields. Here, you have static fields and can use syntax:
```
%MyApp.Content.Event{
  description: ...
}
```

Fln1k committed:
add GraphQL

can request events list with author_email(join user table)

azhi commented on events_task/test/my_app/accounts/accounts_test.exs:
```diff
@@ -6,11 +6,11 @@ defmodule MyApp.AccountsTest do
 
-    @valid_attrs %{email: "some email"}
-    @update_attrs %{email: "some updated email"}
+    @valid_attrs %{email: "sergei@gmail.com"}
```
It's better to test with 'fake' data when possible - by 'fake' I mean data that won't belong to any real person.
This is to protect us from situation when somebody accidentally launched test code in production env - we don't want to actually send an email to `sergei@gmail.com` :).
IANA reserved `example.com` domain for this purpose - it's better to use it in tests.

Fln1k committed:
fix test + upgrade api

now can get all events with author and all confirmed users.

Fln1k committed:
Finish event resolver, start user resolver

Fln1k committed:
Finish GET request part of API

Fln1k committed:
add mutations to API

can create user,event,confirmation

Fln1k committed:
add GraphQL Schema test

Fln1k committed:
add foreign_key errors for confirmation and event

Fln1k committed:
add shema tests

azhi commented on events_task/lib/my_app/accounts/accounts.ex:
```diff
@@ -12,8 +8,15 @@ defmodule MyApp.Accounts do
+  end
+
+  def get_user(param) when is_map(param) do
+    case Enum.at(Map.keys(param), 0) do
```
So if somebody called `get_user(email: "test@example.com", id: 5)`, you will ignore second param and only look by email? This doesn't look like an expected thing :).

It seems that you wanted three different cases to get user by. You can implement it like this:

```
def get_user(id) when is_integer(id), do: Repo.get(User, id)
def get_user(:id, id) when is_integer(id), do: Repo.get(User, id)
def get_user(:email, email) when is_binary(email), do: Repo.get_by(User, email: email)
```

Second function case is not required.

Or, you could give up trying to wrap getting a user, and just pass things to Repo.get_by:
```
def get_user(id) when is_integer(id), do: Repo.get(Repo, id)
def get_user(params) when is_map(params), do: Repo.get_by(Repo, params)
```

Both options would be fine.

azhi commented on events_task/lib/my_app/content/content.ex:
```diff
@@ -48,23 +35,49 @@ defmodule MyApp.Content do
+  def list_confirmations_by_user(id) do
+    from(c in Confirmation,
+      where: c.user_id == ^id,
+      select: c
```
If you are doing `from(c in Confirmation, ...)`, then `select: c` is the default - you can remove it.

azhi commented on events_task/lib/my_app/content/content.ex:
```diff
@@ -48,23 +35,49 @@ defmodule MyApp.Content do
     if user do
-      from(c in Confirmation,
+      from(e in Event,
+        join: c in Confirmation,
```
If you have associations setup in ecto schemas, you can use `assoc` helper instead of joining manually: `join: c in assoc(e, :confirmations)`

azhi commented on events_task/lib/my_app_web/controllers/page_controller.ex:
```diff
@@ -20,7 +20,7 @@ defmodule MyAppWeb.PageController do
     |> assign(
       :current_user_confirmations,
-      Content.list_user_confirmed_event_ids(current_user)
+      for(%{id: id} <- Content.list_events_confirmed(current_user), do: id)
```
Alternative and shorter way of doing this would be
`Content.list_events_confirmed(current_user) |> Enum.map(& &1.id)`

azhi commented on events_task/lib/my_app_web/templates/page/index.html.eex:
```diff
@@ -58,6 +58,7 @@
             <%= if Enum.member?(@current_user_confirmations, event.id) do %>
             Confirmed
             <% else %>
+            <%IO.inspect(@current_user_confirmations) %>
```
^ :)

azhi commented on events_task/test/my_app_web/controllers/confirmation_controller_test.exs:
```diff
@@ -54,20 +52,16 @@ defmodule MyApp.ConfirmationControllerTest do
-               "info" => "Confirmed successfully."
-             }
+      assert get_flash(conn, :error) == nil
+      clear_flash(conn)
```
It doesn't work like that - in elixir you can't mutate a variable inside a function.
`conn = clear_flash(conn)`
But after other changes it looks like you don't need to clear it anymore.

azhi commented on events_task/test/my_app_web/shema_test.exs:
```diff
@@ -0,0 +1,572 @@
+defmodule MyApp.SchemaTest do
```
Typo in file name - `sChema_test.exs`

Fln1k committed:
fix
